File: tmdb-core/src/main/java/dev/reuss/tmdb/domain/trending/model/TrendingMovie.java
package dev.reuss.tmdb.domain.trending.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import dev.reuss.tmdb.common.TmdbModel;
import dev.reuss.tmdb.util.TmdbCollections;
import dev.reuss.tmdb.value.media.MediaType;

import java.util.List;

/**
 * Movie returned by TMDB movie trending.
 *
 * @param adult            whether the movie is marked as adult
 * @param backdropPath     backdrop image path
 * @param id               TMDB movie id
 * @param title            localized movie title
 * @param originalLanguage original language code
 * @param originalTitle    original movie title
 * @param overview         movie overview
 * @param posterPath       poster image path
 * @param mediaType        media type, usually {@code movie}
 * @param genreIds         genre ids
 * @param popularity       movie popularity
 * @param releaseDate      movie release date
 * @param video            whether the movie has video content
 * @param voteAverage      vote average
 * @param voteCount        vote count
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public record TrendingMovie(
        boolean adult,

        @JsonProperty("backdrop_path")
        String backdropPath,

        int id,

        String title,

        @JsonProperty("original_language")
        String originalLanguage,

        @JsonProperty("original_title")
        String originalTitle,

        String overview,

        @JsonProperty("poster_path")
        String posterPath,

        @JsonProperty("media_type")
        MediaType mediaType,

        @JsonProperty("genre_ids")
        List<Integer> genreIds,

        double popularity,

        @JsonProperty("release_date")
        String releaseDate,

        boolean video,

        @JsonProperty("vote_average")
        double voteAverage,

        @JsonProperty("vote_count")
        int voteCount
) implements TmdbModel {

    public TrendingMovie {
        genreIds = TmdbCollections.list(genreIds);
    }
}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/domain/trending/model/TrendingMovie.kt
package dev.reuss.tmdb.domain.trending.model

import com.fasterxml.jackson.annotation.JsonProperty
import dev.reuss.tmdb.common.TmdbModel
import dev.reuss.tmdb.value.media.MediaType

/**
 * Movie returned by TMDB movie trending.
 *
 * @property adult whether the movie is marked as adult
 * @property backdropPath backdrop image path
 * @property id TMDB movie id
 * @property title localized movie title
 * @property originalLanguage original language code
 * @property originalTitle original movie title
 * @property overview movie overview
 * @property posterPath poster image path
 * @property mediaType media type, usually `movie`
 * @property genreIds genre ids
 * @property popularity movie popularity
 * @property releaseDate movie release date
 * @property video whether the movie has video content
 * @property voteAverage vote average
 * @property voteCount vote count
 */
@JvmRecord
data class TrendingMovie(
    @all:JsonProperty("adult")
    val adult: Boolean,

    @all:JsonProperty("backdrop_path")
    val backdropPath: String?,

    @all:JsonProperty("id")
    val id: Int,

    @all:JsonProperty("title")
    val title: String?,

    @all:JsonProperty("original_language")
    val originalLanguage: String?,

    @all:JsonProperty("original_title")
    val originalTitle: String?,

    @all:JsonProperty("overview")
    val overview: String?,

    @all:JsonProperty("poster_path")
    val posterPath: String?,

    @all:JsonProperty("media_type")
    val mediaType: MediaType?,

    @all:JsonProperty("genre_ids")
    val genreIds: List<Int> = emptyList(),

    @all:JsonProperty("popularity")
    val popularity: Double,

    @all:JsonProperty("release_date")
    val releaseDate: String?,

    @all:JsonProperty("video")
    val video: Boolean,

    @all:JsonProperty("vote_average")
    val voteAverage: Double,

    @all:JsonProperty("vote_count")
    val voteCount: Int
) : TmdbModel